test/clustertest/BedrockClusterTester.h

```diff
@@ -1,38 +1,222 @@
 #include <test/lib/BedrockTester.h>
 
-class BedrockClusterTester {
+enum class ClusterSize {
+    ONE_NODE_CLUSTER = 1,
+    THREE_NODE_CLUSTER = 3,
+    FIVE_NODE_CLUSTER = 5,
+    SIX_NODE_CLUSTER = 6,
+};
+
+template <typename T>
+class ClusterTester {
   public:
-    enum ClusterSize {
-        ONE_NODE_CLUSTER   = 1,
-        THREE_NODE_CLUSTER = 3,
-        FIVE_NODE_CLUSTER  = 5,
-        SIX_NODE_CLUSTER  = 6,
-    };
 
     // Creates a cluster of the given size and brings up all the nodes. The nodes will have priority in the order of
     // their creation (i.e., node 0 is highest priority and will become leader.
     // You can also specify plugins to load if for some reason you need to override the default configuration.
-    BedrockClusterTester(ClusterSize size, list<string> queries = {}, int threadID = 0, map<string, string> _args = {}, list<string> uniquePorts = {}, string pluginsToLoad = "db,cache,jobs");
-    BedrockClusterTester(int threadID, string pluginsToLoad = "db,cache,jobs");
-    ~BedrockClusterTester();
+    ClusterTester(ClusterSize size, list<string> queries = {}, int threadID = 0, map<string, string> _args = {}, list<string> uniquePorts = {}, string pluginsToLoad = "db,cache,jobs");
+    ClusterTester(int threadID, string pluginsToLoad = "db,cache,jobs");
+    ClusterTester(const string& pluginString = "db,cache,jobs");
+    ~ClusterTester();
 
-    // Returns the bedrock tester at the given index in the cluster.
-    BedrockTester* getBedrockTester(size_t index);
+    // Returns the tester at the given index in the cluster.
+    T& getTester(size_t index);
+
+    // Legacy name for the above. TODO: Remove when not used by auth.
+    T* getBedrockTester(size_t index);
 
     // Starts a given node, given the same arguments given by the constructor.
-    string startNode(size_t nodeIndex);
+    string startNode(size_t index);
 
     // Same as above but don't wait for the command port to be ready.
-    string startNodeDontWait(size_t nodeIndex);
+    string startNodeDontWait(size_t index);
 
     // Stops a given node.
-    void stopNode(size_t nodeIndex);
+    void stopNode(size_t index);
 
   private:
 
     // The number of nodes in the cluster.
     int _size;
 
     // A list of all our testers that make up our cluster.
-    vector<BedrockTester> _cluster;
+    list<T> _cluster;
 };
+
+typedef ClusterTester<BedrockTester> BedrockClusterTester;
+
+template <typename T>
+ClusterTester<T>::ClusterTester(const string& pluginString) : ClusterTester<T>(
+    ClusterSize::THREE_NODE_CLUSTER,
+    {},
+    0,
+    {},
+    {},
+    pluginString
+    )
+{}
+
+template <typename T>
+ClusterTester<T>::ClusterTester(int threadID, string pluginsToLoad)
+  : ClusterTester<T>(ClusterSize::THREE_NODE_CLUSTER, {},
+                            threadID, {}, {}, pluginsToLoad, nullptr) {}
+
+template <typename T>
+ClusterTester<T>::ClusterTester(ClusterSize size,
```

Comment:
It would be really nice to stick to our/the convention of writing code in `.cpp` files and not `.h` 😬

Reply:
So the reason we can't is because these all use templates:

>Because templates are compiled when required, this forces a restriction for multi-file projects: the implementation (definition) of a template class or function must be in the same file as its declaration. That means that we cannot separate the interface in a separate header file, and that we must include both interface and implementation in any file that uses the templates.


You can read more at the bottom of: http://www.cplusplus.com/doc/oldtutorial/templates/